File: Java/Assignment3/Assignment3.java
package Assignment3.src;

import java.util.Arrays;
import java.util.Collections;
import java.util.Scanner;
import javax.swing.JOptionPane;

public class Assignment3 
{
    static boolean checkbool(boolean[] boo, boolean parameter)
    {
        for (int iter = 0; iter < boo.length; ++iter)
        {
            if (boo[iter] == parameter)
            {
                return true;
            }
        }

        return false;
    }

    static boolean checkint(int[] name, int num, int top)
    {
        for (int iter = name.length - top; iter < name.length; ++iter)
        {
            if (name[iter] == num)
            {
                return true;
            }
        }

        return false;
    }

    static int one(int[][] arr, String dire, int num, int size)
    {
        int sum = 0;

        if (num > size)
        {
            JOptionPane.showMessageDialog(null, "ERROR");
            return -1;
        }

        if (dire.equals("r") || dire.equals("R"))
        {
            for (int i = 0; i < size; i++)
            {
                sum += arr[num][i];
            }
        }
        else if (dire.equals("c") || dire.equals("C"))
        {
            for (int i = 0; i < size; i++)
            {
                sum += arr[i][num];
            }
        }
        else if (dire.equals("fd") || dire.equals("FD"))
        {
            for (int i = 0; i < size; i++)
            {
                sum += arr[num - i][i];
            }
        }
        else if (dire.equals("rd") || dire.equals("RD"))
        {
            for (int i = 0; i < size; i++)
            {
                sum += arr[i][i];
            }
        }

        return sum;
    }

    static String two(int n)
    {
        int dire = 1;
        int direr = 1;
        int count = 1;
        int countr = 1;
        String diamond = "";

        if (n % 2 == 0)
        {
            JOptionPane.showMessageDialog(null, "ERROR");
            return "ERROR";
        }
        else
        {
            int counto = n * 2;

            while (dire > 0)
            {
                if (dire % 2 != 0)
                {
                    if (dire <= n)
                    {
                        diamond += String.join("", Collections.nCopies(counto, " "));
                        diamond+= String.join("", Collections.nCopies(dire, "*"));
                        diamond += "\n";
                        --counto;

                        if (dire == n)
                        {
                            counto += 2;
                        }

                        ++dire;
                    }
                    else if (dire > n)
                    {
                        diamond += String.join("", Collections.nCopies(counto, " "));
                        diamond+= String.join("", Collections.nCopies(dire - 2 * (count + 1), "*"));
                        ++dire;

                        if (dire == 2 * n)
                        {
                            break;
                        }

                        diamond += "\n";
                        count += 2;
                        ++counto;
                    }
                }

                ++dire;
            }

            diamond += "\n\n";

            int countor = n * 2;

            while (direr > 0)
            {
                if (direr % 2 != 0)
                {
                    if (direr <= n)
                    {
                        diamond += String.join("", Collections.nCopies(countor, " "));
                        diamond += "*";

                        if (direr != 1)
                        {
                            diamond += String.join("", Collections.nCopies(direr - 2, " "));
                            diamond += "*";
                        }

                        diamond += "\n";
                        --countor;

                        if (direr == n)
                        {
                            countor += 2;
                        }

                        ++direr;
                    }
                    else if (direr > n)
                    {
                        diamond += String.join("", Collections.nCopies(countor, " "));
                        diamond += "*";

                        if (direr != 2 * n - 1)
                        {
                            diamond += String.join("", Collections.nCopies((direr - 2 * (countr + 1)) - 2, " ")); 
                            diamond += "*";
                        }

                        ++direr;

                        if (direr == 2 * n)
                        {
                            break;
                        }

                        diamond += "\n";
                        countr += 2;
                        ++countor;
                    }
                }

                ++direr;
            }
        }

        return diamond;
    }

    static boolean three(String num)
    {
        String copy = num;
        StringBuilder reverse = new StringBuilder();
        reverse.append(num);
        reverse.reverse();

        if (copy.equals(reverse.toString()))
        {
            return true;
        }
        else
        {
            return false;
        }
    }

    static void four(int lim)
    {
        for (int i = 2; i <= lim; ++i)
        {
            int adf = 0;

            for (int j = 1; j <= i; ++j)
            {
                if (i % j == 0 && j != i)
                {
                    adf += j;
                }
            }

            if (adf == i)
            {
                System.out.println(i + " is a perfect number!");
            }
        }
    }

    static void five(int min, int max)
    {
        boolean[] check = new boolean[max];
        Arrays.fill(check, true);

        while (min < max)
        {
            if (check[min])
            {
                for (int j = min; j != check.length; ++j)
                {
                    if (j % min == 0)
                    {
                        check[j] = false;
                    }
                }
                System.out.println(min + ", ");
            }
            ++min;
        }
    }

    static void six()
    {
        Scanner sc = new Scanner(System.in);
        String password, output = "";
        boolean repeat = true;
        boolean[] rules = new boolean[4];
        Arrays.fill(rules, true);
        int iter;
        String[] build = {"password must be longer than equal to six charachters, ", "passowrd must begin with a capital, ", "password must contain two digits, ", "password must contain one and only one special charachter (!,@,#,$,%,^,&,*)"};
        
        while (repeat)
        {
            System.out.println("Enter a password: ");
            password = sc.nextLine();
            output.replaceAll("", "");

            if (password.length() < 6)
            {
                rules[0] = false;
                output += build[0];
            }

            String glerph = password.substring(0, 1).toUpperCase() + password.substring(1);

            if (password.charAt(0) != glerph.charAt(0)|| Character.isDigit(password.charAt(0)))
            {
                rules[1] = false;
                output += build[1];
            }

            int count = 0;

            for (iter = 0; iter < password.length(); ++iter)
            {
                if (Character.isDigit(password.charAt(iter)))
                {
                    ++count;
                }
            }

            if (count < 2)
            {
                rules[2] = false;
                output += build[2];
            }

            int counto = 0;

            for (int i = 0; i < password.length(); ++i)
            {
                if (password.charAt(i) == '!' || password.charAt(i) == '@' || password.charAt(i) == '#' || password.charAt(i) == '$' || password.charAt(i) == '%' || password.charAt(i) == '^' || password.charAt(i) == '&' || password.charAt(i) == '*')
                {
                    ++counto;
                }
            }

            if (counto != 1)
            {
                rules[3] = false;
                output += build[3];
            }

            if (checkbool(rules, false))
            {
                System.out.println(output);
            }
            else
            {
                System.out.println("Accepted");
                repeat = false;
            }
        }
    }

    static String seven(int[] rain, String[] month)
    {
        String output = "Month\tRainfall Amount\n=====\t===============\n";
        int[] raino = rain.clone();
        Arrays.sort(rain);

        for (int i = 0; i < month.length; ++i)
        {
            output += month[i] + "\t\t";
            output += raino[i];

            if (checkint(rain, raino[i], 3))
            {
                output += "\tBring an umbrella";
            }

            output += "\n";
        }

        return output;
    }

    static void select()
    {
        Scanner sc = new Scanner(System.in);
        System.out.println("1. 2D Addition\n2. Diamond\n3. Palidrome Number\n4. Perfect Numbers\n5. Prime Numbers (Sieve of Eratosthenes)\n6. Password\n7. Rain");
        int choice = sc.nextInt();

        switch (choice)
        {
            case 1:
                int size, num;
                System.out.print("Enter vector size: ");
                size = sc.nextInt();
                int[][] arr = new int[size][size];

                for (int i = 0; i < size; ++i)
                {
                    for (int j = 0; j < size; ++j)
                    {
                        System.out.print("Enter element " + (j + 1) + " in row " + (i + 1) + ": ");
                        arr[i][j] = sc.nextInt();
                    }
                }

                System.out.println("Which section do you want to add? (r, c, fd, rd)");
                String dire = sc.next();
                System.out.println("Which index do you want to check (0 for fd, rd)");
                num = sc.nextInt();
                System.out.println("Return value: " + (one(arr, dire, num, size)));
                return;

            case 2: 
                int n;
                System.out.print("Enter size: ");
                n = sc.nextInt();
                System.out.println(two(n));
                return;

            case 3:
                System.out.print("Enter number: ");
                String numo = sc.next();
                System.out.println(three(numo));
                return;

            case 4:
                System.out.print("Enter limit: ");
                int lim = sc.nextInt();
                four(lim);
                return;

            case 5:
                System.out.println("Enter minimum: ");
                int min = sc.nextInt();
                System.out.println("Enter maximum: ");
                int max = sc.nextInt();
                five(min, max);
                return;

            case 6:
                six();
                return;

            case 7:
                String[] month = {"Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"};
                int[] rain = new int[12];

                for (int i = 0; i < month.length; ++i)
                {
                    System.out.println("Enter rainfall amount for " + month[i] + ": ");
                    rain[i] = sc.nextInt();
                }
            
                System.out.println(seven(rain, month));
                return;

            default:
                JOptionPane.showMessageDialog(null, "ERROR");
                return;
        }
    }
    public static void main(String[] args) throws Exception 
    {
        while (true)
        {
            select();
        }
    }
}

File: C++/Assignment3/Assignment3.cpp
#include <iostream>
#include <vector>
#include <string>
#include <windows.h>
#include <algorithm>

using namespace std;

bool checkbool(vector<bool> boo, bool parameter)
{
    for (auto iter = boo.begin(); iter < boo.end(); ++iter)
    {
        if (*iter == parameter)
        {
            return true;
        }
    }

    return false;
}

int getInt()
{
    int temp;
    cin >> temp;
    cin.ignore();
    return temp;
}

bool checkint(vector<int> name, int num, int top)
{
    for (auto iter = name.end() - top; iter < name.end(); ++iter)
    {
        if (*iter == num)
        {
            return true;
        }
    }

    return false;
}

int one(vector<vector<int>> vec, string choose, int num, int size)
{
    int sum = 0;

    if (num > size)
    {
        MessageBox(nullptr, L"ERROR", L"ERROR", MB_OK);
        return -1;
    }

    if (choose == "r" || choose == "R")
    {
        for (size_t i = 0; i < size; i++)
        {
            sum += vec[num][i];
        }
    }
    else if (choose == "c" || choose == "C")
    {
        for (size_t i = 0; i < size; i++)
        {
            sum += vec[i][num];
        }
    }
    else if (choose == "fd" || choose == "FD")
    {
        for (size_t i = 0; i < size; i++)
        {
            sum += vec[num - i][i];
        }
    }
    else if (choose == "rd" || choose == "RD")
    {
        for (size_t i = 0; i < size; i++)
        {
            sum += vec[i][i];
        }
    }

    return sum;
}

string two(int n)
{
    int dire = 1;
    int direr = 1;
    int count = 1;
    int countr = 1;
    string diamond = "";

    if (n % 2 == 0)
    {
        MessageBox(nullptr, L"ERROR", L"ERROR", MB_OK);
        return "ERROR";
    }
    else
    {
        int counto = n * 2;

        while (dire > 0)
        {
            if (dire % 2 != 0)
            {
                if (dire <= n)
                {
                    diamond.insert(diamond.end(), counto, ' ');
                    diamond.insert(diamond.end(), dire, '*');
                    diamond += "\n";
                    --counto;

                    if (dire == n)
                    {
                        counto += 2;
                    }

                    ++dire;
                }
                else if (dire > n)
                {
                    diamond.insert(diamond.end(), counto, ' ');
                    diamond.insert(diamond.end(), dire - 2 * (count + 1), '*');
                    ++dire;

                    if (dire == 2 * n)
                    {
                        break;
                    }

                    diamond += "\n";
                    count += 2;
                    ++counto;
                }
            }

            ++dire;
        }

        diamond += "\n\n";

        int countor = n * 2;

        while (direr > 0)
        {
            if (direr % 2 != 0)
            {
                if (direr <= n)
                {
                    diamond.insert(diamond.end(), countor, ' ');
                    diamond += "*";

                    if (direr != 1)
                    {
                        diamond.insert(diamond.end(), direr - 2, ' ');
                        diamond += "*";
                    }

                    diamond += "\n";
                    --countor;

                    if (direr == n)
                    {
                        countor += 2;
                    }

                    ++direr;
                }
                else if (direr > n)
                {
                    diamond.insert(diamond.end(), countor, ' ');
                    diamond += "*";

                    if (direr != 2 * n - 1)
                    {
                        diamond.insert(diamond.end(), (direr - 2 * (countr + 1)) - 2, ' ');
                        diamond += "*";
                    }

                    ++direr;

                    if (direr == 2 * n)
                    {
                        break;
                    }

                    diamond += "\n";
                    countr += 2;
                    ++countor;
                }
            }

            ++direr;
        }
    }

    return diamond;
}

bool three(string num)
{
    string copy = num;
    reverse(num.begin(), num.end());

    if (copy == num)
    {
        return true;
    }
    else
    {
        return false;
    }
}

void four(int lim)
{
    for (unsigned int i = 2; i <= lim; ++i)
    {
        int adf = 0;

        for (unsigned int j = 1; j <= i; ++j)
        {
            if (i % j == 0 && j != i)
            {
                adf += j;
            }
        }

        if (adf == i)
        {
            cout << i << " is a perfect number!" << endl;
        }
    }
}

void five(int min, int max)
{
    vector<bool> check(max, true);

    while (min < max)
    {
        if (check[min])
        {
            for (int j = min; j != check.size(); ++j)
            {
                if (j % min == 0)
                {
                    check[j] = false;
                }
            }
            cout << min << ", ";
        }
        ++min;
    }

    cout << endl;
}

void six()
{
    bool (*checkBool)(vector<bool>, bool) = &checkbool;
    string password, output;
    bool repeat = true;
    vector<bool> rules(4, true);
    string::iterator iter;
    vector<string> build{"password must be longer than equal to six charachters, ", "passowrd must begin with a capital, ", "password must contain two digits, ", "password must contain one and only one special charachter (!,@,#,$,%,^,&,*)"};
    
    while (repeat)
    {
        cout << "Enter a password: ";
        cin >> password;
        cin.ignore();
        output.clear();

        if (password.size() < 6)
        {
            rules[0] = false;
            output += build[0];
        }

        if (password.front() != toupper(password.front()) || isdigit(password.front()))
        {
            rules[1] = false;
            output += build[1];
        }

        int count = 0;

        for (iter = password.begin(); iter < password.end(); ++iter)
        {
            if (isdigit(*iter))
            {
                ++count;
            }
        }

        if (count < 2)
        {
            rules[2] = false;
            output += build[2];
        }

        int counto = 0;

        for (auto i = password.begin(); i < password.end(); ++i)
        {
            if (*i == '!' || *i == '@' || *i == '#' || *i == '$' || *i == '%' || *i == '^' || *i == '&' || *i == '*')
            {
                ++counto;
            }
        }

        if (counto != 1)
        {
            rules[3] = false;
            output += build[3];
        }

        if ((*checkBool)(rules, false))
        {
            cout << output << endl;
        }
        else
        {
            cout << "Accepted" << endl;
            repeat = false;
        }
    }
}

string seven(vector<int> rain, vector<string> month)
{
    bool (*checkInt)(vector<int>, int, int) = &checkint;
    string output = "Month\tRainfall Amount\n=====\t===============\n";
    vector<int> raino = rain;
    nth_element(rain.begin(), rain.begin() + 2, rain.end());

    for (size_t i = 0; i < month.size(); ++i)
    {
        output += month[i] + "\t\t";
        output += to_string(raino[i]);

        if ((*checkInt)(rain, raino[i], 3))
        {
            output += "\tBring an umbrella";
        }

        output += "\n";
    }

    return output;
}

void select()
{
    int (*getint)() = &getInt;
    int (*ones)(vector<vector<int>>, string, int, int) = &one;
    string (*twos)(int) = &two;
    bool (*threes)(string) = &three;
    void (*fours)(int) = &four;
    void (*fives)(int, int) = &five;
    void (*sixes)() = &six;
    string (*sevens)(vector<int>, vector<string>) = &seven;
    vector<vector<int>> vec;
    vector<int> rain;
    vector<string> month{"Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"};
    int choice;
    string choose;
    string numo;
    cout << "1. 2D addition\n2. Diamond\n3. Palindrome\n4. Perfect Numbers\n5. Prime Numbers(Sieve of Eratosthenes)\n6. Password\n7. Rain" << endl;
    cin >> choice;
    cin.ignore();

    switch (choice)
    {
        case 1:
            unsigned int size, num;
            cout << "Enter vector size: ";
            cin >> size;
            cin.ignore();
            vec.resize(size, vector<int>(size));

            for (size_t i = 0; i < size; ++i)
            {
                for (size_t j = 0; j < size; ++j)
                {
                    cout << "Enter element " << (j + 1) << " in row " << (i + 1) << ": ";
                    cin >> vec[i][j];
                    cin.ignore();
                }
            }

            cout << "Which section do you want to add? (r, c, fd, rd)" << endl;
            cin >> choose;
            cin.ignore();
            cout << "Which index do you want to check (0 for fd, rd)" << endl;
            cin >> num;
            cin.ignore();
            cout << "Return value: " << to_string((*ones)(vec, choose, num, size)) << endl;
            return;

        case 2: 
            unsigned int n;
            cout << "Enter size: ";
            cin >> n;
            cin.ignore();
            cout << (*twos)(n) << endl;
            return;

        case 3:
            cout << "Enter number: ";
            cin >> numo;
            cin.ignore();
            cout << boolalpha << (*threes)(numo) << endl;
            return;

        case 4:
            int lim;
            cout << "Enter limit: ";
            cin >> lim;
            cin.ignore();
            (*fours)(lim);
            return;

        case 5:
            int min, max;
            cout << "Enter minimum: ";
            cin >> min;
            cin.ignore();
            cout << "Enter maximum: ";
            cin >> max;
            cin.ignore();
            (*fives)(min, max);
            return;

        case 6:
            (*sixes)();
            return;

        case 7:

            for (size_t i = 0; i < month.size(); ++i)
            {
                cout << "Enter rainfall amount for " << month[i] << ": ";
                rain.push_back((*getint)());
            }

            cout << (*sevens)(rain, month) << endl;
            return;

        default:
            MessageBox(nullptr, L"ERROR", L"ERROR", MB_OK);
            return;
    }
}

int main()
{
    void (*sele)() = &select;

    while (true)
    {
        (*sele)();
    }
}